File: scripts/generate-causal-parity.py
"""Freeze independently computed observations for explicitly causal contracts."""
import json, math, subprocess, sys
from pathlib import Path
from parity_sources import verify_source
ROOT=Path(__file__).resolve().parents[1]
verify_source('ta');verify_source('ttr')
sys.path.insert(0,str(ROOT/'target/cross-library/ta'))
import pandas as pd
from ta.trend import KSTIndicator
data=json.loads((ROOT/'crates/roze-ta/tests/fixtures/python-parity.json').read_text())
records=[]
def clean(x): return float(x) if math.isfinite(x) else None
for f in data['fixtures']:
    if (f['source'],f['name'])!=('ta','KSTIndicator'):continue
    bars=data['datasets'][f['dataset']]
    close=pd.Series([s['value'] for s in bars])
    p=f['request']['operation']['params']
    kwargs={**{f'roc{i}':p[f'roc{i}'] for i in range(1,5)},**{f'window{i}':p[f'sma{i}'] for i in range(1,5)},'nsig':p['signal']}
    f['expected']=[]
    for size in range(1,len(close)+1):
        obj=KSTIndicator(close=close.iloc[:size],**kwargs)
        f['expected'].append({'kst':clean(obj.kst().iloc[-1]),'signal':clean(obj.kst_sig().iloc[-1])})
    f['alignment']='prefix_snapshot_last; early ROC fill mean uses only observed prefix'
    records.append(f)
subprocess.run(['rtk','proxy',str(ROOT/'target/R-runtime/app/bin/x64/Rscript.exe'),'scripts/generate-causal-parity.R'],cwd=ROOT,check=True)
records+=json.loads((ROOT/'target/causal-ttr.json').read_text())
(ROOT/'crates/roze-ta/tests/fixtures/causal-parity.json').write_text(json.dumps({'schema_version':1,'fixtures':records},indent=2)+'\n',encoding='utf8',newline='\n')
print(f'Generated {len(records)} independent causal fixtures')

File: scripts/generate-causal-parity.R
# Independent prefix observations: never execute the Rust implementation.
.libPaths(c(normalizePath('target/R-libraries'),.libPaths()))
suppressPackageStartupMessages(library(TTR))
oracle <- new.env(parent=asNamespace('TTR'))
for (file in list.files('target/cross-library/ttr/R',pattern='[.]R$',full.names=TRUE)) sys.source(file,oracle)
dll <- dyn.load(normalizePath('target/cross-library/ttr-msvc/build/Release/TTR.dll'))
routines <- getDLLRegisteredRoutines(dll)$.Call
for (name in names(routines)) assign(paste0('C_',name),routines[[name]],oracle)
data <- jsonlite::fromJSON('crates/roze-ta/tests/fixtures/ttr-parity.json',simplifyVector=FALSE)
records <- list()
for (f in data$fixtures) {
  if (!(f$name %in% c('ZigZag','DPO'))) next
  bars <- data$datasets[[f$dataset]]
  hl <- do.call(rbind,lapply(bars,function(s) unlist(s$value[c('high','low')])))
  if(f$name=='ZigZag') {
    # The native oracle reads two initial bars: one-bar input is explicitly warming.
    f$expected <- lapply(seq_len(nrow(hl)),function(i) if(i<2) NA_real_ else tail(oracle$ZigZag(hl[seq_len(i),,drop=FALSE],change=f$request$operation$params$threshold*100),1))
    f$alignment <- 'prefix_snapshot_last; one sample is warming (reference requires two)'
  } else {
    delay <- floor(f$request$operation$params$period/2)+1
    f$expected <- c(rep(list(NA_real_),delay),head(f$expected,-delay))
    f$alignment <- paste0('delayed_by_',delay,'_observations; source value t emitted at t+delay')
  }
  records[[length(records)+1]] <- f
}
jsonlite::write_json(records,'target/causal-ttr.json',auto_unbox=TRUE,digits=NA,na='null',null='null')
